// @sanity/preview-kit/dist/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
var react = require('react');
var isFastEqual = require('react-fast-compare');
var withSelector = require('use-sync-external-store/with-selector');
var context = require('./_chunks/context-JrKnN2Wb.cjs');
var jsxRuntime = require('react/jsx-runtime');
function _interopDefaultCompat(e) {
  return e && typeof e === 'object' && 'default' in e ? e : {
    default: e
  };
}
var isFastEqual__default = /*#__PURE__*/_interopDefaultCompat(isFastEqual);
function useLiveQuery(initialData, query, queryParams, options) {
  const {
    isEqual = isFastEqual__default.default
  } = options || {};
  const defineStore = react.useContext(context.defineListenerContext);
  const params = context.useParams(queryParams);
  const store = react.useMemo(() => defineStore(initialData, query, params), [defineStore, initialData, params, query]);
  const [serverSnapshot] = react.useState(() => {
    if (initialData === void 0) {
      throw new Error("initialSnapshot can't be undefined, if you don't want an initial value use null instead");
    }
    try {
      return JSON.parse(JSON.stringify(initialData));
    } catch (error) {
      console.warn("Failed to deep clone initialSnapshot, this is likely an error and an indication that the snapshot isn't JSON serializable", {
        initialSnapshot: initialData,
        error
      });
      return initialData;
    }
  });
  const getServerSnapshot = react.useCallback(() => serverSnapshot, [serverSnapshot]);
  const selector = react.useCallback(snapshot2 => snapshot2, []);
  const snapshot = withSelector.useSyncExternalStoreWithSelector(store.subscribe, store.getSnapshot, getServerSnapshot, selector, isEqual);
  const loading = useLiveQueryIsLoading(query, params);
  return [snapshot, loading];
}
function useLiveQueryIsLoading(query, params) {
  const loadedListeners = react.useContext(context.LoadedListenersContext);
  const key = react.useMemo(() => context.getQueryCacheKey(query, params), [params, query]);
  return react.useMemo(() => {
    if (Array.isArray(loadedListeners)) {
      return loadedListeners.includes(key) ? false : true;
    }
    return false;
  }, [key, loadedListeners]);
}
function useIsEnabled() {
  return react.useContext(context.IsEnabledContext);
}
function createLiveQueryProvider(options) {
  const {
    GroqStoreProvider,
    LiveStoreProvider
  } = options;
  const SelectStoreProvider = react.memo(props => {
    var _a;
    const {
      children,
      refreshInterval,
      client,
      cache,
      logger
    } = props;
    const maxDocuments = (_a = cache == null ? void 0 : cache.maxDocuments) != null ? _a : context.DEFAULT_MAX_DOCUMENTS;
    const [documentsCount, setDocumentsCount] = react.useState(null);
    const [error, setError] = react.useState(null);
    const [includeTypes] = react.useState(() => {
      var _a2;
      return (_a2 = cache == null ? void 0 : cache.includeTypes) != null ? _a2 : [];
    });
    if (error) {
      throw error;
    }
    react.useEffect(() => {
      if (documentsCount !== null) {
        return;
      }
      logger == null ? void 0 : logger.log("[@sanity/preview-kit]: fetch documents count");
      const controller = new AbortController();
      client.fetch(includeTypes.length > 0 ? "count(*[_type in $includeTypes])" : "count(*)", {
        includeTypes
      }, {
        filterResponse: true,
        signal: controller.signal
      }).then(result => {
        logger == null ? void 0 : logger.log("[@sanity/preview-kit]: documents count", result);
        setDocumentsCount(result);
      }).catch(error2 => {
        if (error2.name !== "AbortError") {
          setError(error2);
        }
      });
      return () => {
        controller.abort();
      };
    }, [client, documentsCount, includeTypes, logger]);
    if (documentsCount === null) {
      return children;
    }
    if (refreshInterval && documentsCount >= maxDocuments) {
      return /* @__PURE__ */jsxRuntime.jsx(react.Suspense, {
        fallback: children,
        children: /* @__PURE__ */jsxRuntime.jsx(LiveStoreProvider, {
          client,
          refreshInterval,
          turboSourceMap: false,
          logger,
          children
        })
      });
    }
    if (documentsCount >= maxDocuments) {
      throw new Error("[@sanity/preview-kit]: You have ".concat(documentsCount, " documents in your dataset, which is more than the maximum of ").concat(maxDocuments, " documents. Please reduce the number of documents in your dataset or increase the limit."));
    }
    return /* @__PURE__ */jsxRuntime.jsx(react.Suspense, {
      fallback: children,
      children: /* @__PURE__ */jsxRuntime.jsx(GroqStoreProvider, {
        ...props,
        children
      })
    });
  });
  SelectStoreProvider.displayName = "SelectStoreProvider";
  const LiveQueryProvider = react.memo(props => {
    const {
      children,
      refreshInterval = 1e4,
      token
    } = props;
    if (!props.client) {
      throw new Error("Missing a `client` prop with a configured Sanity client instance");
    }
    const [client] = react.useState(() => {
      const {
        requestTagPrefix,
        resultSourceMap
      } = props.client.config();
      return props.client.withConfig({
        requestTagPrefix: requestTagPrefix || context.DEFAULT_TAG,
        resultSourceMap: resultSourceMap || "withKeyArraySelector",
        // Set the recommended defaults, this is a convenience to make it easier to share a client config from a server component to the client component
        ...(token && {
          token,
          useCdn: false,
          perspective: "previewDrafts",
          ignoreBrowserTokenWarning: true
        })
      });
    });
    const [cache] = react.useState(() => props.cache);
    const [logger] = react.useState(() => props.logger);
    const turboSourceMap = react.useMemo(() => {
      var _a;
      const {
        resultSourceMap
      } = client.config();
      return (_a = props.turboSourceMap) != null ? _a : resultSourceMap === "withKeyArraySelector" || resultSourceMap;
    }, [client, props.turboSourceMap]);
    if (turboSourceMap) {
      return /* @__PURE__ */jsxRuntime.jsx(react.Suspense, {
        fallback: children,
        children: /* @__PURE__ */jsxRuntime.jsx(LiveStoreProvider, {
          client,
          logger,
          refreshInterval,
          turboSourceMap,
          children
        })
      });
    }
    return /* @__PURE__ */jsxRuntime.jsx(react.Suspense, {
      fallback: children,
      children: /* @__PURE__ */jsxRuntime.jsx(SelectStoreProvider, {
        client,
        cache,
        logger,
        refreshInterval,
        children
      })
    });
  });
  LiveQueryProvider.displayName = "LiveQueryProvider";
  return LiveQueryProvider;
}
const GroqStoreProvider = react.lazy(() => Promise.resolve().then(function () {
  return require('./_chunks/index-_eh5eWPW.cjs');
}));
const LiveStoreProvider = react.lazy(() => Promise.resolve().then(function () {
  return require('./_chunks/index-FcsTxES2.cjs');
}));
const LiveQueryProvider = createLiveQueryProvider({
  GroqStoreProvider,
  LiveStoreProvider
});
exports.LiveQueryProvider = LiveQueryProvider;
exports.createLiveQueryProvider = createLiveQueryProvider;
exports.default = LiveQueryProvider;
exports.useIsEnabled = useIsEnabled;
exports.useLiveQuery = useLiveQuery;
