// sanity/lib/desk.esm.js
import { C as ComponentBuilder$1, a as ComponentViewBuilder$1, b as ConfirmDeleteDialogContainer, D as DEFAULT_INTENT_HANDLER$1, S as StructureToolProvider, c as DocumentBuilder$1, d as DocumentInspectorHeader$1, e as DocumentListBuilder$1, f as DocumentListItemBuilder$1, g as DocumentListPane$1, h as DocumentPane$1, i as DocumentPaneProvider$1, j as DocumentTypeListBuilder$1, F as FormViewBuilder$1, G as GenericListBuilder$1, k as GenericViewBuilder$1, H as HELP_URL$1, I as InitialValueTemplateItemBuilder$1, L as ListBuilder$1, l as ListItemBuilder$1, M as MenuItemBuilder$1, m as MenuItemGroupBuilder$1, P as PaneLayout$1, n as PaneRouterContext$1, o as SerializeError$1, p as component$1, q as createStructureBuilder$1, r as defaultInitialValueTemplateItems$1, s as defaultIntentChecker$1, t as structureTool, u as documentFromEditor$1, v as documentFromEditorWithInitialValue$1, w as form$1, x as getOrderingMenuItem$1, y as getOrderingMenuItemsForSchemaType$1, z as getTypeNamesFromFilter$1, A as isDocumentListItem$1, B as maybeSerializeInitialValueTemplateItem$1, E as maybeSerializeMenuItem$1, J as maybeSerializeMenuItemGroup$1, K as maybeSerializeView$1, N as menuItemsFromInitialValueTemplateItems$1, O as shallowIntentChecker$1, Q as structureLocaleNamespace$1, R as useStructureTool, T as useDocumentPane$1, U as useDocumentTitle$1, V as usePaneRouter$1 } from "./_legacy/pane.esm.js";
import "./_legacy/generate-help-url.esm.esm.js";
const ComponentBuilder = ComponentBuilder$1, ComponentViewBuilder = ComponentViewBuilder$1, ConfirmDeleteDialog = ConfirmDeleteDialogContainer, DEFAULT_INTENT_HANDLER = DEFAULT_INTENT_HANDLER$1, DeskToolProvider = StructureToolProvider, DocumentBuilder = DocumentBuilder$1, DocumentInspectorHeader = DocumentInspectorHeader$1, DocumentListBuilder = DocumentListBuilder$1, DocumentListItemBuilder = DocumentListItemBuilder$1, DocumentListPane = DocumentListPane$1, DocumentPane = DocumentPane$1, DocumentPaneProvider = DocumentPaneProvider$1, DocumentTypeListBuilder = DocumentTypeListBuilder$1, FormViewBuilder = FormViewBuilder$1, GenericListBuilder = GenericListBuilder$1, GenericViewBuilder = GenericViewBuilder$1, HELP_URL = HELP_URL$1, InitialValueTemplateItemBuilder = InitialValueTemplateItemBuilder$1, ListBuilder = ListBuilder$1, ListItemBuilder = ListItemBuilder$1, MenuItemBuilder = MenuItemBuilder$1, MenuItemGroupBuilder = MenuItemGroupBuilder$1, PaneLayout = PaneLayout$1, PaneRouterContext = PaneRouterContext$1, SerializeError = SerializeError$1, component = component$1, createStructureBuilder = createStructureBuilder$1, defaultInitialValueTemplateItems = defaultInitialValueTemplateItems$1, defaultIntentChecker = defaultIntentChecker$1, deskTool = structureTool, documentFromEditor = documentFromEditor$1, documentFromEditorWithInitialValue = documentFromEditorWithInitialValue$1, form = form$1, getOrderingMenuItem = getOrderingMenuItem$1, getOrderingMenuItemsForSchemaType = getOrderingMenuItemsForSchemaType$1, getTypeNamesFromFilter = getTypeNamesFromFilter$1, isDocumentListItem = isDocumentListItem$1, maybeSerializeInitialValueTemplateItem = maybeSerializeInitialValueTemplateItem$1, maybeSerializeMenuItem = maybeSerializeMenuItem$1, maybeSerializeMenuItemGroup = maybeSerializeMenuItemGroup$1, maybeSerializeView = maybeSerializeView$1, menuItemsFromInitialValueTemplateItems = menuItemsFromInitialValueTemplateItems$1, shallowIntentChecker = shallowIntentChecker$1, structureLocaleNamespace = structureLocaleNamespace$1, useDeskTool = useStructureTool, useDocumentPane = useDocumentPane$1, useDocumentTitle = useDocumentTitle$1, usePaneRouter = usePaneRouter$1;
export {
  ComponentBuilder,
  ComponentViewBuilder,
  ConfirmDeleteDialog,
  DEFAULT_INTENT_HANDLER,
  DeskToolProvider,
  DocumentBuilder,
  DocumentInspectorHeader,
  DocumentListBuilder,
  DocumentListItemBuilder,
  DocumentListPane,
  DocumentPane,
  DocumentPaneProvider,
  DocumentTypeListBuilder,
  FormViewBuilder,
  GenericListBuilder,
  GenericViewBuilder,
  HELP_URL,
  InitialValueTemplateItemBuilder,
  ListBuilder,
  ListItemBuilder,
  MenuItemBuilder,
  MenuItemGroupBuilder,
  PaneLayout,
  PaneRouterContext,
  SerializeError,
  component,
  createStructureBuilder,
  defaultInitialValueTemplateItems,
  defaultIntentChecker,
  deskTool,
  documentFromEditor,
  documentFromEditorWithInitialValue,
  form,
  getOrderingMenuItem,
  getOrderingMenuItemsForSchemaType,
  getTypeNamesFromFilter,
  isDocumentListItem,
  maybeSerializeInitialValueTemplateItem,
  maybeSerializeMenuItem,
  maybeSerializeMenuItemGroup,
  maybeSerializeView,
  menuItemsFromInitialValueTemplateItems,
  shallowIntentChecker,
  structureLocaleNamespace,
  useDeskTool,
  useDocumentPane,
  useDocumentTitle,
  usePaneRouter
};
